app/internal/security/ouath2.py

```diff
@@ -0,0 +1,127 @@
+            return LoginUser(
+                username=user.username, password=db_user.password)
+    return False
+
+
+def create_jwt_token(
+        user: LoginUser, JWT_MIN_EXP: int = JWT_MIN_EXP,
+        JWT_KEY: str = JWT_KEY) -> str:
+    """Creating jwt-token out of user unique data"""
+    expiration = datetime.utcnow() + timedelta(minutes=JWT_MIN_EXP)
+    jwt_payload = {
+        "sub": user.username,
+        "hashed_password": user.password,
+        "exp": expiration}
+    jwt_token = jwt.encode(
+        jwt_payload, JWT_KEY, algorithm=JWT_ALGORITHM)
+    return jwt_token
+
+
+async def check_jwt_token(
+    db: Session,
+    token: str = Depends(oauth_schema),
+        path: bool = None) -> CurrentUser:
+    """
+    Check whether JWT token is correct. Returns User object if yes.
+    Returns None or raises HTTPException,
+    depanding which depandency activated this function.
+    """
+    try:
+        jwt_payload = jwt.decode(
+            token, JWT_KEY, algorithms=JWT_ALGORITHM)
+        jwt_username = jwt_payload.get("sub")
+        jwt_hashed_password = jwt_payload.get("hashed_password")
+        db_user = await get_db_user_by_username(db, username=jwt_username)
+        if db_user and db_user.password == jwt_hashed_password:
+            return CurrentUser(**db_user.__dict__)
+        else:
+            raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                headers=path,
+                detail="Your token is incorrect. Please log in again")
+    except InvalidSignatureError:
+        raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                headers=path,
+                detail="Your token is incorrect. Please log in again")
+    except jwt.ExpiredSignatureError:
+        raise HTTPException(
+            status_code=HTTP_401_UNAUTHORIZED,
+            headers=path,
+            detail="Your token has expired. Please log in again")
+    except jwt.DecodeError:
+        raise HTTPException(
+            status_code=HTTP_401_UNAUTHORIZED,
+            headers=path,
+            detail="Your token is incorrect. Please log in again")
+
+
+async def get_cookie(request: Request) -> str:
+    """
+    Extracts jwt from HTTPONLY cookie, if exists.
+    Raises HTTPException if not.
+    """
+    if 'Authorization' in request.cookies:
+        return request.cookies['Authorization']
+    raise HTTPException(
+        status_code=HTTP_401_UNAUTHORIZED,
+        headers=request.url.path,
+        detail="Please log in to enter this page")
+
+
+async def my_exception_handler(
+        request: Request,
+        exc: HTTP_401_UNAUTHORIZED) -> RedirectResponse:
+    """
+    Whenever HTTP_401_UNAUTHORIZED is raised,
+    redirecting to login route, with original requested url,
+    and details for why original request failed.
+    """
+    paramas = f"?next={exc.headers}&message={exc.detail}"
+    url = f"/login{paramas}"
```

Comment:
Please use url_for

Reply:
https://www.starlette.io/routing/
For the reason i am not in a route, but in a regular function, i don't have the router instance over there.
Using request.url_for is possible, but because i want to pass query parameters, it is much more complicated.
In flask it's much more simple..
After reading allot, i really think that the way i wrote it is much more simple, then trying to pass parameters the same way the documentation suggests.
Can i leave this one as it is?

Reply:
I have tried to do it. url_for uses router, which i don't have here and i have to import it. i have tried to do so, but i got stuck with circular importing that i could not solve. I also tried to move this function to login.py, where i have the router, but still got circular importing, this time with testing. i was unable to resolve it, and i hope we can just leave this one like that.. :)

Reply:
:+1: Got you